### PR TITLE
Dev/stam/accounts

### PR DESCRIPTION
**Pull request by psarras**
This seems like it is going what it is supposed to. However, due to an existing bug, I am a bit uncertain. 
If you orchestrated everything manually instead of relying on the auto refresh, it works. Otherwise, it requires the sender to trigger twice. I think this is a problem only if you have the sender and the receiving component of the same stream on one GH canvas...

Let me know if I am missing something. The sender / receiver component are doing way more than they should so it is hard to have confidence on them, ergo the "Draft PR"

### CODE REVIEW

**Comment by HughGrovesArup** [comment]
@psarras - what's the context of this PR? What should the user be able to do after it is implemented?

**Comment by psarras** [comment]
@HughGrovesArup The user would be able to optionally add an Account input, using the context menu on the sender component. This would be used to validate the send. As I understand it this is important for Rhino Compute. Hope this makes sense.

**Comment by HughGrovesArup** [comment]
I cant seem to see that behaviour? What do you see?

image

**Comment by psarras** [comment]
I meant receiver...sorry

File: SpeckleGrasshopper/BaseComponents/GhReceiverCoreClient.cs [hunk 1]
```diff
@@ -258,6 +267,35 @@ public override void AppendAdditionalMenuItems( ToolStripDropDown menu )
...
+    {
+      AccountRequired = !AccountRequired;
+      
+      if(AccountRequired)
+      {
+        Params.Input.Insert(1, new Param_GenericObject() 
+        {
+          Name = "Account",
+          NickName = "Ac", 
+          Access = GH_ParamAccess.item,
```
Comment (HughGrovesArup) [inline SpeckleGrasshopper/BaseComponents/GhReceiverCoreClient.cs:284]: Need to add a description in as well

File: SpeckleGrasshopper/Management/CreateAccount.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+namespace SpeckleGrasshopper.Management
+{
+  public class CreateAccount : GH_Component
+  {
+    /// <summary>
+    /// Initializes a new instance of the CreateAccount class.
+    /// </summary>
+    public CreateAccount()
+      : base("CreateAccount", "CA",
```
Comment (HughGrovesArup) [inline SpeckleGrasshopper/Management/CreateAccount.cs:15]: I'm probably out of the loop on this component, but I don't really understand its purpose. Is it really niche? Should it be hidden from regular users if its something only to support TDA?

Also it needs an icon.

File: SpeckleGrasshopper/Management/CreateAccount.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+          "Speckle", "Management")
+    {
+    }
+
+    /// <summary>
+    /// Registers all the input parameters for this component.
+    /// </summary>
+    protected override void RegisterInputParams(GH_InputParamManager pManager)
+    {
+      pManager.AddIntegerParameter("AccountId", "id", "AccountId", GH_ParamAccess.item);
```
Comment (HughGrovesArup) [inline SpeckleGrasshopper/Management/CreateAccount.cs:26]: The long descriptions are no more helpful than the parameter names

**Review by HughGrovesArup — CHANGES_REQUESTED** [review]
I've not really tested the actual behaviour of these changes as I'm not sure how. Is it to ensure that you are using a specific server when pulling a stream?

File: SpeckleGrasshopper/Management/CreateAccount.cs (hunk 2 above)
Comment (psarras) [reply SpeckleGrasshopper/Management/CreateAccount.cs:15]: Good points. @daviddekoning thoughts?

Comment (psarras) [reply SpeckleGrasshopper/Management/CreateAccount.cs:15]: We should probably wait on this PR to do the reorganisation, I guess so we can give it a proper place. As I understand it @HughGrovesArup  it is a bit niche. It would allow you to set the account, without communicating with other servers, provided you have all the credentails.

**Comment by psarras** [comment]
@HughGrovesArup I am using this script.

[testAccounts.zip](https://github.com/arup-group/SpeckleRhino/files/5408222/testAccounts.zip)

File: SpeckleGrasshopper/Management/CreateAccount.cs (hunk 2 above)
Comment (daviddekoning) [reply SpeckleGrasshopper/Management/CreateAccount.cs:15]: It should be under the 'Advanced' sub-section, but not otherwise hidden. Those are the components that most users should ignore. 

Yes, we'll need an icon. For the GetApplicationID, I just whipped something up in Gimp, but we can also contact a graphic designer within the firm somewhere to give us something. If you guys have a contact, please let me know, otherwise I'll reach out to some folks.

Comment (daviddekoning) [reply SpeckleGrasshopper/Management/CreateAccount.cs:15]: @psarras, @HughGrovesArup - yes, this work is not for this week's release

**Comment by daviddekoning** [comment]
@HughGrovesArup, this is to allow scripts to run in TDA / rhino.inside, where a single script can be run many times by many people. I'll update things in Jira - there are a few other bits we need to think through as well (e.g. how do we control when a sender creates a new stream vs. updates on old one)

File: SpeckleGrasshopper/BaseComponents/GhReceiverCoreClient.cs (hunk 1 above)
Comment (psarras) [reply SpeckleGrasshopper/BaseComponents/GhReceiverCoreClient.cs:285]: Amended this one

File: SpeckleGrasshopper/Management/CreateAccount.cs (hunk 3 above)
Comment (psarras) [reply SpeckleGrasshopper/Management/CreateAccount.cs:27]: tried to add a bit more nuance. I think some of them are not needed actually. Through testing it seems only the "Token" + the "Api" are the important ones. I wonder if we just remove the other ones... @daviddekoning any thoughts?

File: SpeckleGrasshopper/Management/CreateAccount.cs (hunk 2 above)
Comment (psarras) [reply SpeckleGrasshopper/Management/CreateAccount.cs:16]: @daviddekoning oops, just saw your comment regarding the icon...argh I've made a quick one as well ...

Comment (daviddekoning) [reply SpeckleGrasshopper/Management/CreateAccount.cs:16]: I ❤️ quick icons

File: SpeckleGrasshopper/Management/CreateAccount.cs (hunk 3 above)
Comment (daviddekoning) [reply SpeckleGrasshopper/Management/CreateAccount.cs:27]: Yea, I think you are right that those are the only important ones. Given that our aim is to write a component that supports the TDA workflow, let's just expose those two. If we need the other info, the component can pull it from the server via the account API.

**Review by daviddekoning — CHANGES_REQUESTED** [review]
hi @psarras  - this looks good. I've tested it with receivers and I am able to add a receiver to a canvas for one server and then specify an account on another server.

Let's keep going with the other changes we discussed (global setting for manually specifying the account, and adding the ability to specify the account for a sender) and then merge the whole thing

**Comment by psarras** [comment]
@daviddekoning I added a toggle on the Create Account component that allows to make this a global Account. Also now both sender and receiver have the option to add the Account. Let me know what you think.

**Comment by daviddekoning** [comment]
By global account, do you mean that you add the account to the Speckle database and it is accessible to all programs on that computer?

**Comment by daviddekoning** [comment]
We should also be able to specify the stream ID on the sender, but only when it goes into manual mode.

**Comment by psarras** [comment]
> By global account, do you mean that you add the account to the Speckle database and it is accessible to all programs on that computer?

Nothing so fancy. The sender / receiver look at a static variable if this flag is on. 

> We should also be able to specify the stream ID on the sender, but only when it goes into manual mode.

So the Account iD is the thing that gets specified the stream id, is taken care of from the Client class.
Currently the logic goes: If you have global flag on, then you follow this Account, else you follow the one provided as an input, else you follow the one that is set from the GUI.

**Comment by daviddekoning** [comment]
I have tested this and it appears to work nicely. There are a couple of rough edges, but I think we should publish a version with this functionality so that the TDA team can start working, then polish it up a bit.

I have pushed a commit with some name changes.

Here are the items I noticed - none of which are showstoppers:

1. If a `Receiver` does not have a StreamId attached, and the global toggle is switched, the `Account` input does not appear. The input appears when you attach a StreamId. 
2. When the mode is switched back to "regular", the account selector popup appears. It would be really nice if the components would remember what the previous account was when they go to Rhino.Compute mode, so that they can continue behaving as they did before if it is switched off. That being said - this is a good behaviour for now.
3. When Rhino.Compute mode is on, if you drop a new Receiver onto the canvas, it still asks you to specify the Account (related to item 1 above).
4. It would be great if the Speckle Menu item `Toggle Rhino.Compute mode` had a checkmark to reflect the current mode.

In summary - let's get this merged into the the main branch and create a release so that the TDA team can start putting it through it's paces. We will then update it based on their feedback, possibly incorporating the above points.

**Review by daviddekoning — APPROVED** [review]

**Comment by psarras** [comment]
@daviddekoning 

I think 1), should be and easy fix. 
2, 3 are related and they will be a bit much for this PR although definitely doable. Was this the previous behaviour? If I broke it, I would like to bring this back!

4) I think there should be a way. I will have a look.

BTW the stam/dev/sychronous is quite a small change, maybe we can include it here? I think it is important for TDA.

**Comment by daviddekoning** [comment]
Thanks @psarras. I dot think 2&3 are regressions, just new behaviours that emerged when we made the changes.

Yes, please include the synchronous changes here - then all the Rhino.Compute functionality are included in one place. 

One last thing, can you set up some tests to make sure that the components work reliably with Rhino.compute mode off. I don't mind if the new functionality is a bit flaky, but we should not compromise the old behaviour.

**Comment by daviddekoning** [comment]
Hi @psarras, the overall UX behaviour is good, however, there seems to be some bugginess in sending/receiving data in Rhino.Compute mode.

1. Receivers sometimes need to be force refreshed after being switched into R.C mode. To reproduce:
    1. Add a receiver to a canvas
    2. Toggle R.C mode to "On"
    3. Connect a streamId and account
    4. Data may preview in the Rhino window, but is not output from the output parameter
    5. Force refresh the component, and all is good
2. Senders aren't always sending all data when the inputs update.

**Comment by psarras** [comment]
@daviddekoning 

I think the 1) / 2) are existing behaviours that I raised in the past? Generally on the Sender there is a debouncer that is preventing some of the inputs to be send. On the receiver I have noticed that because this works with events, it is a bit precarious if while you are updating you are receiving data as well.

I will double check on 1) If there is any funny business specific to switching th R.C on and get back to you!
